pkg/controller/statefulset/stateful_set_control_test.go

```diff
@@ -1072,6 +1072,153 @@ func TestStatefulSetControlOnDeleteUpdate(t *testing.T) {
+		if err != nil {
+			t.Fatalf("%s: %s", test.name, err)
+		}
+		pods, err := spc.podsLister.Pods(set.Namespace).List(selector)
+		if err != nil {
+			t.Fatalf("%s: %s", test.name, err)
+		}
+		set, err = spc.setsLister.StatefulSets(set.Namespace).Get(set.Name)
+		if err != nil {
+			t.Fatalf("%s: %s", test.name, err)
+		}
+		if err := test.validate(set, pods); err != nil {
+			t.Fatalf("%s: %s", test.name, err)
+		}
+	}
+
+	originalImage := newStatefulSet(3).Spec.Template.Spec.Containers[0].Image
+
+	tests := []testcase{
+		{
+			name:       "monotonic image update",
+			paused:     false,
+			invariants: assertMonotonicInvariants,
+			initial: func() *appsv1alpha1.StatefulSet {
+				return newStatefulSet(3)
+			},
+			update: func(set *appsv1alpha1.StatefulSet) *appsv1alpha1.StatefulSet {
+				set.Spec.Template.Spec.Containers[0].Image = "foo"
+				return set
+			},
+			validate: func(set *appsv1alpha1.StatefulSet, pods []*v1.Pod) error {
+				sort.Sort(ascendingOrdinal(pods))
+				for i := range pods {
+					if pods[i].Spec.Containers[0].Image != "foo" {
+						return fmt.Errorf("want pod %s image foo found %s", pods[i].Name, pods[i].Spec.Containers[0].Image)
+					}
+				}
+				return nil
+			},
+		},
+		{
+			name:       "monotonic image update with paused",
+			paused:     true,
+			invariants: assertMonotonicInvariants,
+			initial: func() *appsv1alpha1.StatefulSet {
+				return newStatefulSet(3)
+			},
+			update: func(set *appsv1alpha1.StatefulSet) *appsv1alpha1.StatefulSet {
+				set.Spec.Template.Spec.Containers[0].Image = "foo"
+				return set
+			},
+			validate: func(set *appsv1alpha1.StatefulSet, pods []*v1.Pod) error {
+				sort.Sort(ascendingOrdinal(pods))
+				for i := range pods {
+					if pods[i].Spec.Containers[0].Image != originalImage {
+						return fmt.Errorf("want pod %s image %s found %s", pods[i].Name, originalImage, pods[i].Spec.Containers[0].Image)
+					}
+				}
+				return nil
+			},
+		},
+		{
+			name:       "burst image update",
+			paused:     false,
+			invariants: assertBurstInvariants,
+			initial: func() *appsv1alpha1.StatefulSet {
+				return burst(newStatefulSet(3))
+			},
+			update: func(set *appsv1alpha1.StatefulSet) *appsv1alpha1.StatefulSet {
+				set.Spec.Template.Spec.Containers[0].Image = "foo"
+				return set
+			},
+			validate: func(set *appsv1alpha1.StatefulSet, pods []*v1.Pod) error {
+				sort.Sort(ascendingOrdinal(pods))
+				for i := range pods {
+					if pods[i].Spec.Containers[0].Image != "foo" {
+						return fmt.Errorf("want pod %s image foo found %s", pods[i].Name, pods[i].Spec.Containers[0].Image)
+					}
+				}
+				return nil
+			},
+		},
+		{
+			name:       "burst image update with paused",
+			paused:     true,
+			invariants: assertBurstInvariants,
+			initial: func() *appsv1alpha1.StatefulSet {
+				return burst(newStatefulSet(3))
+			},
+			update: func(set *appsv1alpha1.StatefulSet) *appsv1alpha1.StatefulSet {
+				set.Spec.Template.Spec.Containers[0].Image = "foo"
```

Comment:
can we also increase replicas to 4 besides update images, so that we can validate that an new pod is created but existing pods is not updated